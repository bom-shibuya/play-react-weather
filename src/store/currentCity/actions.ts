export const SWITCH_CITY = 'SWITCH_CITY'

export const switchCity = (payload: { city: string }) => ({
  payload,
  type: SWITCH_CITY as typeof SWITCH_CITY
})

export type Actions = ReturnType<typeof switchCity>
