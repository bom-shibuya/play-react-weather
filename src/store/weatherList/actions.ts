import { WeatherList } from './types'

export const FETCH_REQUEST = 'FETCH_REQUEST'
export const FETCH_SUCCESS = 'FETCH_SUCCESS'
export const FETCH_FAILURE = 'FETCH_FAILURE'

export const fetchRequest = (payload: { city: string }) => ({
  payload,
  type: FETCH_REQUEST as typeof FETCH_REQUEST
})

export interface IFetchSuccessPayload {
  city: string
  data: WeatherList
}

export const fetchSuccess = (payload: IFetchSuccessPayload) => ({
  payload,
  type: FETCH_SUCCESS as typeof FETCH_SUCCESS
})

export const fetchFailure = (payload: { city: string }) => ({
  payload,
  type: FETCH_FAILURE as typeof FETCH_FAILURE
})

export type Actions =
  | ReturnType<typeof fetchRequest>
  | ReturnType<typeof fetchSuccess>
  | ReturnType<typeof fetchFailure>
