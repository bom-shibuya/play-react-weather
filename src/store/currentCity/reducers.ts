import { Actions, SWITCH_CITY } from './actions'

export interface ICurrentCityState {
  currentCity: string
}

export const initialState: ICurrentCityState = {
  currentCity: 'Tokyo'
}

export const currentCity = (state = initialState, action: Actions) => {
  switch (action.type) {
    case SWITCH_CITY:
      return { ...state, currentCity: action.payload.city }
    default:
      return state
  }
}
