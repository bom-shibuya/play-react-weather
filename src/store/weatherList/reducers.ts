import { WeatherList } from '../../types'
import { Actions, FETCH_FAILURE, FETCH_REQUEST, FETCH_SUCCESS } from './actions'

export const initialState = {
  isError: false,
  isFetching: false,
  lastUpdateAt: null as null | number,
  list: [] as WeatherList
}

export type WeathersState = typeof initialState

export const weathers = (
  state = initialState,
  action: Actions
): WeathersState => {
  switch (action.type) {
    case FETCH_REQUEST:
      return { ...state, isFetching: true }
    case FETCH_SUCCESS:
      return {
        isError: false,
        isFetching: false,
        lastUpdateAt: Date.now(),
        list: [...action.payload.data]
      }
    case FETCH_FAILURE:
      return {
        ...state,
        isError: true,
        lastUpdateAt: Date.now()
      }
    default:
      return state
  }
}

export interface IWetherListState {
  [key: string]: WeathersState
}

export const weatherList = (state = {}, action: Actions): IWetherListState => {
  switch (action.type) {
    case FETCH_REQUEST:
    case FETCH_SUCCESS:
    case FETCH_FAILURE:
      return {
        ...state,
        [action.payload.city]: weathers(initialState, action)
      }
    default:
      return state
  }
}
