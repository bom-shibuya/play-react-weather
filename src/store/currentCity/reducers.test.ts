import snapshotDiff from 'snapshot-diff'
import { switchCity } from './actions'
import { currentCity, initialState } from './reducers'

describe('currentCity reducer test', () => {
  test('Reducer return initial state', () => {
    expect(
      snapshotDiff(initialState, currentCity(initialState, {} as any))
    ).toMatchSnapshot()
  })

  test('Reducer return state currentCity => "Osaka"', () => {
    expect(
      snapshotDiff(
        initialState,
        currentCity(initialState, switchCity({ city: 'Osaka' }))
      )
    ).toMatchSnapshot()
  })
})
