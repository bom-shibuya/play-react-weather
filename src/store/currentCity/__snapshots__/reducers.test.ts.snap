// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currentCity reducer test Reducer return initial state 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`currentCity reducer test Reducer return state currentCity => "Osaka" 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
-   \\"currentCity\\": \\"Tokyo\\",
+   \\"currentCity\\": \\"Osaka\\",
  }"
`;
